### PR TITLE
Update constructor stubbing

### LINKED ISSUE #60: Doubling a constructor leaks to other test file
I have following test in a file, which worked fine

    def test_valid_message_checkr(self):
     """Test lock works and locked message is not submitted"""

    allow(module).is_enabled_city.and_return(False)

    consumer = InstanceDouble('bgc.lib.kafka.Consumer')
    with no_builtin_verification():
      allow(module.Consumer).__new__.and_return(consumer)

      allow(consumer).get_envelopes.and_return(
          [ujson.dumps(self.kafka_bgc_log)],
      )

      allow(consumer).lock_message.and_return(False)

      expect(application).submit.never()

      module.main()

Somehow this results in totally unrelated failure in other file

_______________________________________________________ TestConsumer.test_get_envelopes _______________________________________________________
Traceback (most recent call last):

    File "/home/uber/background-check/env/local/lib/python2.7/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
    File "/home/uber/background-check/bgc/tests/lib/kafka/kafka_consumer_test.py", line 26, in   test_get_envelopes
    consumer = Consumer('background-check.submit', 42)
    DeprecationWarning: object.new() takes no parameters

### PR DESCRIPTION
**Pull request by toddsifleet**
* Let users stub constructors on class doubles with allow_constructor and expect_constructor.
* Add patch and patch_class, similar to mock.patch
* Don't force users to patch __new__ to control new instances of classes that are created. 
* Will allow us to stop doing allow(Class).__new__, which will resolve #60

### CODE REVIEW

**Comment by toddsifleet** [comment]
@breerly @charlax @UberFarmer @lucasmarshall @samongoose I would love some feedback on this PR before I merge it in.

File: docs/source/faq.rst [hunk 1]
```diff
@@ -18,6 +34,61 @@ Python does not allow doubles to look into builtin functions and asked them what
...
+
+Use ``patch_class`` and ``allow_constructor``::
+
+    from doubles import patch_class, allow_constructor
+
+    import myapp
+
+    def test_something_that_uses_user():
+        patched_class = patch_class('myapp.user.User')
+        allow_constructor(patch_class).and_return('Bob Barker')
```
Comment (charlax) [inline docs/source/faq.rst]: `patched_class`? I'd recommend including test files inside the doc to have tested documentation.

File: docs/source/usage.rst [hunk 2]
```diff
@@ -381,3 +382,60 @@ If you ever want to to clear all allowances and expectations you have set on an
...
+
+    from doubles import patch
+    import doubles.testing
+
+    def test_patch():
+        patch('doubles.testing.User', 'Bob Barker')
+
+        assert doubles.testing.User == 'Bob Barker'
+
+Patches do not verify against the underlying object, so use them carefully.  Patches are automatically restored at the end of the test.
```
Comment (charlax) [inline docs/source/usage.rst]: What do you think about clarifying why this is not a good idea?

File: docs/source/usage.rst [hunk 3]
```diff
@@ -381,3 +382,60 @@ If you ever want to to clear all allowances and expectations you have set on an
...
+    import doubles.testing
+
+    def test_allow_constructor_with_args():
+        User = ClassDouble('doubles.testing.User')
+
+        allow_constructor(User).with_args('Bob', 100).and_return('Bob')
+
+        assert User('Bob', 100) == 'Bob'
+
+The return value of ``allow_constructor`` and ``expect_constructor`` support all of the same methods as allow/expect. (e.g. ``with_args``, ``once``, ``exactly``, .etc).  As with ``expect``, ``expect_constructor`` does not support ``and_return``, ``and_return_result_of``, or ``and_raise``.
```
Comment (charlax) [inline docs/source/usage.rst]: What about linking to `expect`'s documentation, making it clearly explain why it does not support those?

File: doubles/class_double.py [hunk 4]
```diff
@@ -13,3 +34,27 @@ class ClassDouble(InstanceDouble):
...
+        self._doubles_target = patch_class(self._doubles_target)
+        self._target = Target(self._doubles_target)
+
+    def __call__(self, *args, **kwargs):
+        """
+        represents __new__ for ``ClassDouble`` instances.
+
+        :rtype obj:
+        :raises VerifyingDoubleArgumentError: If args/kwargs don't match the expected argumenst of
+        __init__ of the underlying class.
```
Comment (charlax) [inline doubles/class_double.py]: Might want to indent this.

File: doubles/class_double.py [hunk 5]
```diff
@@ -13,3 +34,27 @@ class ClassDouble(InstanceDouble):
...
     is_class = True
+
+    def __init__(self, path):
+        super(ClassDouble, self).__init__(path)
+        self._doubles_target = patch_class(self._doubles_target)
+        self._target = Target(self._doubles_target)
+
+    def __call__(self, *args, **kwargs):
+        """
+        represents __new__ for ``ClassDouble`` instances.
```
Comment (charlax) [inline doubles/class_double.py]: First line of a docstring should a one-liner. http://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

Nit: docstring should prescribe the function or method's effect as a command ("Return this"), not as a description ("Returns that") :) http://www.python.org/dev/peps/pep-0257/

File: doubles/patch.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        :param str path: The absolute module path to the class.
+        """
+        module_path, self._name = get_path_components(target)
+        self.target = get_module(module_path, target)
+        self._capture_original_object()
+        self.set_value(None)
+
+    def _capture_original_object(self):
+        """
+        Capture the original pythong object.
```
Comment (charlax) [inline doubles/patch.py]: typo

**Comment by charlax** [comment]
Looks good! I mostly looked at the documentation.

File: doubles/class_double.py (hunk 5 above)
Comment (toddsifleet) [inline doubles/class_double.py]: I know I have just tried to stay consistent with all of the other docstrings.

File: docs/source/faq.rst (hunk 1 above)
Comment (toddsifleet) [inline docs/source/faq.rst]: I will do this in a different diff.

File: docs/source/usage.rst (hunk 3 above)
Comment (toddsifleet) [inline docs/source/usage.rst]: done

**Comment by lucasmarshall** [comment]
This looks good to me!

**Comment by UberFarmer** [comment]
I'm happy with the new functionality. Didn't look at implementation.

**Comment by breerly** [comment]
+1

**Comment by toddsifleet** [comment]
Squash merged onto master

**Pull request closed without merging**